File: aoc2024/rust/src/util/direction.rs
#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Hash)]
pub enum Direction {
    North,
    NorthWest,
    West,
    SouthWest,
    South,
    SouthEast,
    East,
    NorthEast,
}

impl std::fmt::Display for Direction {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::North => "north".to_string(),
            Self::South => "south".to_string(),
            Self::East => "east".to_string(),
            Self::West => "west".to_string(),
            Self::NorthEast => "northeast".to_string(),
            Self::NorthWest => "northwest".to_string(),
            Self::SouthEast => "southeast".to_string(),
            Self::SouthWest => "southwest".to_string(),
        };

        write!(f, "{s}")
    }
}

File: aoc2024/zig/src/lib/direction.zig
const std = @import("std");

pub const Direction = enum {
    north,
    northwest,
    west,
    southwest,
    south,
    southeast,
    east,
    northeast,

    /// Converts the direction into a string
    pub fn to_string(self: Direction) []const u8 {
        return switch (self) {
            .north => "north",
            .south => "south",
            .east => "east",
            .west => "west",
            .northeast => "northeast",
            .northwest => "northwest",
            .southeast => "southeast",
            .southwest => "southwest",
        };
    }

    /// Debug print the direction with a trailing newline
    pub fn display(self: Direction) !void {
        std.debug.print("{s}\n", .{self.to_string()});
    }
};

test "direction to_string" {
    try std.testing.expectEqualStrings("north", Direction.north.to_string());
    try std.testing.expectEqualStrings("south", Direction.south.to_string());
    try std.testing.expectEqualStrings("east", Direction.east.to_string());
    try std.testing.expectEqualStrings("west", Direction.west.to_string());
    try std.testing.expectEqualStrings("northeast", Direction.northeast.to_string());
    try std.testing.expectEqualStrings("northwest", Direction.northwest.to_string());
    try std.testing.expectEqualStrings("southeast", Direction.southeast.to_string());
    try std.testing.expectEqualStrings("southwest", Direction.southwest.to_string());
}
